### PR TITLE
Add molecule encodings for contextual bandit

### PR DESCRIPTION
**Pull request by YamLyubov**
Adds molecular encoders for contextual graphs. 
Implemented approaches:
- `ECPF` - Extended-Connectivity Fingerprints -
 https://www.semanticscholar.org/paper/Extended-Connectivity-Fingerprints-Rogers-Hahn/d1b796ff0c1895426de11a1eaafc5443be29645d
- `RDKF` - RDKit topological fingerprint - 
 https://www.rdkit.org/docs/GettingStartedInPython.html#rdkit-topological-fingerprints
- `atom_pair` - atom pair fingerprint - https://pubs.acs.org/doi/10.1021/ci00046a002
- `topoligical_torsion` - topoligical torsion fingerprint - https://pubs.acs.org/doi/10.1021/ci00054a008
- `mol_descriptors` - molecular descriptors - 
 https://www.semanticscholar.org/paper/Molecular-representation-learning-with-language-and-Fabian-Edlich/e4c5e81e6e337bb94af3eb719df5f029b40434fa
- `Mol2Vec` - vector representations (pretrained model) - 
 https://pubs.acs.org/doi/10.1021/acs.jcim.7b00616
- `MoleculeTransformer` - vector representation (pretrained model) -  
 https://pdfs.semanticscholar.org/2b80/1b911a5a33a4fe523b667c5f6b1b8bbadbbf.pdf 

Collected materials: https://docs.google.com/spreadsheets/d/1qDx6hhRMeA185_fLxBRwwcHHmvUOxbG9ID_IkOYZV-g/edit?usp=sharing

Results for 10 trials:
qed score
Figure_1

normalized sa score
Figure_2
Figure_3

#176

### CODE REVIEW

**Review by maypink — CHANGES_REQUESTED** [review]

File: examples/molecule_search/mol_encoders.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    def __init__(self):
+        self.file_path = os.path.join(project_root(), Mol2Vec.PRETRAINED_WORD2VEC)
+        download_from_github(self.file_path,
+                             Mol2Vec.GITHUB_URL,
+                             message="Downloading pretrained model for molecules encoding...")
+
+        self.model = word2vec.Word2Vec.load(self.file_path)
+
+    @adapter_method_to_molgraph
+    def __call__(self, obs):
```
[the remark this answers is not in the document]
Comment (YamLyubov) [reply examples/molecule_search/mol_encoders.py]: [Done](https://github.com/aimclub/GOLEM/pull/179/files#diff-71802992c43fe0e311d87484347702e172dc9cc2dc06708e00aad6de2eecfa5bR130)

File: examples/molecule_search/mol_encoders.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        molecule = obs.get_rw_molecule()
+        sentence = MolSentence(mol2alt_sentence(molecule, radius=1))
+        embedding = self.sentences2vec([sentence], self.model)[0]
+        embedding = ensure_wrapped_in_sequence(embedding)
+        if len(embedding) < 300:
```
[the remark this answers is not in the document]
Comment (YamLyubov) [reply examples/molecule_search/mol_encoders.py]: Это было нужно, так как если модели не было известно слово, она возвращала 0. Сейчас сделала иначе - [неизвестные слова модель будет обрабатывать как "UNK"](https://github.com/aimclub/GOLEM/pull/179/files#diff-71802992c43fe0e311d87484347702e172dc9cc2dc06708e00aad6de2eecfa5bR132)

File: examples/molecule_search/mol_encoders.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+    def __init__(self, embedding_size=512, num_layers=6, max_length=256):
+        self.log = default_log(self)
+
+        self.file_path = os.path.join(project_root(), MoleculeTransformer.PRETRAINED_TRANSFORMER)
+        download_from_github(self.file_path,
+                             MoleculeTransformer.GITHUB_URL,
+                             message="Downloading pretrained model for molecules encoding...")
+
+        model = Transformer(ALPHABET_SIZE, embedding_size, num_layers).eval()
```
[the remark this answers is not in the document]
Comment (YamLyubov) [reply examples/molecule_search/mol_encoders.py:190]: [Done](https://github.com/aimclub/GOLEM/pull/179/files#diff-71802992c43fe0e311d87484347702e172dc9cc2dc06708e00aad6de2eecfa5bR189)

File: examples/molecule_search/mol_transformer/transformer.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+class Transformer(nn.Module):
+    def __init__(self, alphabet_size, d_model, n, heads=8, dropout=0.1):
+        super().__init__()
+        self.encoder = Encoder(alphabet_size, d_model, n, heads, dropout)
+        self.decoder = Decoder(alphabet_size, d_model, n, heads, dropout)
+        self.out = nn.Linear(d_model, alphabet_size)
+
+    def forward(self, src, trg, src_mask, trg_mask):
+        e_outputs = self.encoder(src, src_mask)
+        # print("DECODER")
```
[the remark this answers is not in the document]
Comment (YamLyubov) [reply examples/molecule_search/mol_transformer/transformer.py]: [Done](https://github.com/aimclub/GOLEM/pull/179/files#diff-cea8159037dc9fa211095d850573df5148e5adf89a8fa05ba274c859711e8bdcR120-R123)

File: test/unit/adaptive/test_context_agents.py [hunk 5]
```diff
@@ -31,4 +32,4 @@ def test_operations_encoding(context_agent_enum, result_encoding):
     repo = ContextAgentsRepository()
     context_agent = repo.agent_class_by_id(context_agent_enum)
     encoding = context_agent(obs=graph, available_operations=['1', '2', '3', '4', '5'])
-    assert encoding == result_encoding
+    assert np.all(encoding == result_encoding)
```
[the remark this answers is not in the document]
Comment (YamLyubov) [reply test/unit/adaptive/test_context_agents.py]: Done

**Review by maypink — APPROVED** [review]
